File: src/api/decoders/live_student_course_landing.ts
// Strict decoding for the minimal current Student Course Landing projection.

import type {
  LiveStudentAssessmentLandingSummary,
  AssessmentGradeContribution,
  LiveStudentCourseInvitationSummary,
  LiveStudentCourseLandingSummary,
  StudentCourseActiveAttempt,
  StudentCourseProgressAssessment,
  StudentLatestFeedback,
} from "../live_student_course_landing";
import type { AssessmentAttemptCompletion } from "../../../generated/api/AssessmentAttemptCompletion";
import { ASSESSMENT_TYPE_VALUES } from "../../../generated/api/AssessmentType";
import {
  DecodeError,
  decodeArray,
  decodeBoolean,
  decodeFiniteNumber,
  decodeNonnegativeInteger,
  decodeNullable,
  decodePositiveInteger,
  decodeRecord,
  decodeStringEnum,
  decodeString,
} from "../decoder";
import {
  decodeAssessmentId,
  decodeAssessmentTitle,
  decodeCourseInstanceId,
  decodeCourseName,
  field,
  requireOnlyFields,
} from "./shared";
import { decodeStudentAssessmentDecision } from "./student_assessment_decision";
import { decodeCourseTerm } from "./course_term";
import { parseAssessmentAttemptId } from "../../navigation/public_route";

const ASSESSMENT_ATTEMPT_COMPLETIONS = [
  "inProgress",
  "completed",
] as const satisfies ReadonlyArray<AssessmentAttemptCompletion>;

function decodeCourseSummary(value: unknown, path: string): LiveStudentCourseLandingSummary {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["id", "shortName", "longName"]);
  return {
    id: decodeCourseInstanceId(field(record, "id", path), `${path}.id`),
    shortName: decodeCourseName(field(record, "shortName", path), `${path}.shortName`),
    longName: decodeCourseName(field(record, "longName", path), `${path}.longName`),
  };
}

function decodeInvitationSummary(value: unknown, path: string): LiveStudentCourseInvitationSummary {
  const record = decodeRecord(value, path);
  // ASVS 1.5.2/8.2.3: this closed projection contains no private identity fields.
  requireOnlyFields(record, path, ["id", "shortName", "longName", "instructorDisplayName", "term"]);
  const instructorDisplayName = decodeString(
    field(record, "instructorDisplayName", path),
    `${path}.instructorDisplayName`,
  );
  if (
    instructorDisplayName !== instructorDisplayName.trim() ||
    Array.from(instructorDisplayName).length === 0 ||
    Array.from(instructorDisplayName).length > 200 ||
    /[\p{Cc}]/u.test(instructorDisplayName)
  ) {
    throw new DecodeError(`${path}.instructorDisplayName`, "one verified Instructor display name");
  }
  return {
    id: decodeCourseInstanceId(field(record, "id", path), `${path}.id`),
    shortName: decodeCourseName(field(record, "shortName", path), `${path}.shortName`),
    longName: decodeCourseName(field(record, "longName", path), `${path}.longName`),
    instructorDisplayName,
    term: decodeCourseTerm(field(record, "term", path), `${path}.term`),
  };
}

function decodeAssessmentGradeContribution(
  value: unknown,
  path: string,
): AssessmentGradeContribution {
  // ASVS 1.5.1/2.2.1: allow-list the complete finite pair at the browser boundary.
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["pointsEarned", "pointsPossible"]);
  const pointsEarned = decodeFiniteNumber(
    field(record, "pointsEarned", path),
    `${path}.pointsEarned`,
  );
  const pointsPossible = decodeFiniteNumber(
    field(record, "pointsPossible", path),
    `${path}.pointsPossible`,
  );
  if (pointsEarned < 0 || pointsPossible < 0) {
    throw new DecodeError(path, "a complete nonnegative Assessment grade contribution");
  }
  return { pointsEarned, pointsPossible };
}

function decodeAssessmentSummary(
  value: unknown,
  path: string,
): LiveStudentAssessmentLandingSummary {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "id",
    "title",
    "assessmentType",
    "decision",
    "assessmentAttemptNumber",
    "assessmentAttemptCompletion",
    "canResumeAssessmentAttempt",
    "gradedQuestionCount",
    "savedQuestionCount",
    "questionCount",
    "assessmentScore",
  ]);
  const assessmentAttemptNumber = decodeNullable(
    field(record, "assessmentAttemptNumber", path),
    `${path}.assessmentAttemptNumber`,
    decodePositiveInteger,
  );
  const assessmentAttemptCompletion = decodeNullable(
    field(record, "assessmentAttemptCompletion", path),
    `${path}.assessmentAttemptCompletion`,
    (candidate, candidatePath) =>
      decodeStringEnum(candidate, candidatePath, ASSESSMENT_ATTEMPT_COMPLETIONS),
  );
  const gradedQuestionCount = decodeNonnegativeInteger(
    field(record, "gradedQuestionCount", path),
    `${path}.gradedQuestionCount`,
  );
  const questionCount = decodePositiveInteger(
    field(record, "questionCount", path),
    `${path}.questionCount`,
  );
  // ASVS 2.2.1/2.2.3: saved progress is bounded by the retained issued total.
  const savedQuestionCount = decodeNonnegativeInteger(
    field(record, "savedQuestionCount", path),
    `${path}.savedQuestionCount`,
  );
  const assessmentScoreValue = record.assessmentScore;
  const assessmentScore =
    assessmentScoreValue === undefined
      ? undefined
      : decodeAssessmentGradeContribution(assessmentScoreValue, `${path}.assessmentScore`);
  if (
    gradedQuestionCount > questionCount ||
    savedQuestionCount > questionCount ||
    (assessmentAttemptCompletion === null &&
      (assessmentAttemptNumber !== null ||
        gradedQuestionCount !== 0 ||
        savedQuestionCount !== 0)) ||
    (assessmentAttemptCompletion !== null && assessmentAttemptNumber === null)
  ) {
    throw new DecodeError(path, "internally consistent self-only Assessment progress");
  }
  return {
    id: decodeAssessmentId(field(record, "id", path), `${path}.id`),
    title: decodeAssessmentTitle(field(record, "title", path), `${path}.title`),
    assessmentType: decodeStringEnum(
      field(record, "assessmentType", path),
      `${path}.assessmentType`,
      ASSESSMENT_TYPE_VALUES,
    ),
    decision: decodeStudentAssessmentDecision(field(record, "decision", path), `${path}.decision`),
    assessmentAttemptNumber,
    assessmentAttemptCompletion,
    canResumeAssessmentAttempt: decodeBoolean(
      field(record, "canResumeAssessmentAttempt", path),
      `${path}.canResumeAssessmentAttempt`,
    ),
    gradedQuestionCount,
    savedQuestionCount,
    questionCount,
    ...(assessmentScore === undefined ? {} : { assessmentScore }),
  };
}

function decodeProgressAssessment(value: unknown, path: string): StudentCourseProgressAssessment {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "id",
    "title",
    "assessmentType",
    "assessmentAttemptCount",
    "submittedAssessmentAttemptCount",
    "latestAssessmentAttemptNumber",
    "latestAssessmentAttemptCompletion",
    "latestActivityAt",
    "assessmentScore",
    "assessmentScoreIsLatestAttempt",
  ]);
  const assessmentAttemptCount = decodeNonnegativeInteger(
    field(record, "assessmentAttemptCount", path),
    `${path}.assessmentAttemptCount`,
  );
  const submittedAssessmentAttemptCount = decodeNonnegativeInteger(
    field(record, "submittedAssessmentAttemptCount", path),
    `${path}.submittedAssessmentAttemptCount`,
  );
  const latestAssessmentAttemptNumber = decodeNullable(
    field(record, "latestAssessmentAttemptNumber", path),
    `${path}.latestAssessmentAttemptNumber`,
    decodePositiveInteger,
  );
  const latestAssessmentAttemptCompletion = decodeNullable(
    field(record, "latestAssessmentAttemptCompletion", path),
    `${path}.latestAssessmentAttemptCompletion`,
    (candidate, candidatePath) =>
      decodeStringEnum(candidate, candidatePath, ASSESSMENT_ATTEMPT_COMPLETIONS),
  );
  const latestActivityAt = decodeNullable(
    field(record, "latestActivityAt", path),
    `${path}.latestActivityAt`,
    decodeNonnegativeInteger,
  );
  const assessmentScore =
    record.assessmentScore === undefined
      ? undefined
      : decodeAssessmentGradeContribution(record.assessmentScore, `${path}.assessmentScore`);
  const assessmentScoreIsLatestAttempt =
    record.assessmentScoreIsLatestAttempt === undefined
      ? undefined
      : decodeBoolean(
          record.assessmentScoreIsLatestAttempt,
          `${path}.assessmentScoreIsLatestAttempt`,
        );
  if (
    submittedAssessmentAttemptCount > assessmentAttemptCount ||
    (assessmentAttemptCount === 0 &&
      (latestAssessmentAttemptNumber !== null ||
        latestAssessmentAttemptCompletion !== null ||
        latestActivityAt !== null ||
        submittedAssessmentAttemptCount !== 0 ||
        assessmentScore !== undefined ||
        assessmentScoreIsLatestAttempt !== undefined)) ||
    (assessmentAttemptCount > 0 &&
      (latestAssessmentAttemptNumber === null ||
        latestAssessmentAttemptNumber > assessmentAttemptCount ||
        latestAssessmentAttemptCompletion === null ||
        latestActivityAt === null)) ||
    (assessmentScore !== undefined && submittedAssessmentAttemptCount === 0) ||
    (assessmentScore !== undefined && assessmentScoreIsLatestAttempt === undefined) ||
    (assessmentScore === undefined && assessmentScoreIsLatestAttempt !== undefined)
  ) {
    throw new DecodeError(path, "internally consistent self-only Course Progress");
  }
  return {
    id: decodeAssessmentId(field(record, "id", path), `${path}.id`),
    title: decodeAssessmentTitle(field(record, "title", path), `${path}.title`),
    assessmentType: decodeStringEnum(
      field(record, "assessmentType", path),
      `${path}.assessmentType`,
      ASSESSMENT_TYPE_VALUES,
    ),
    assessmentAttemptCount,
    submittedAssessmentAttemptCount,
    latestAssessmentAttemptNumber,
    latestAssessmentAttemptCompletion,
    latestActivityAt,
    ...(assessmentScore === undefined ? {} : { assessmentScore }),
    ...(assessmentScoreIsLatestAttempt === undefined ? {} : { assessmentScoreIsLatestAttempt }),
  };
}

/** Rejects anything beyond the current Student's minimal Course landing projection. */
export function decodeLiveStudentCourseLandings(
  value: unknown,
  path = "response",
): ReadonlyArray<LiveStudentCourseLandingSummary> {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["courses"]);
  return decodeArray(field(record, "courses", path), `${path}.courses`, decodeCourseSummary);
}

/** Rejects anything beyond the current Student's minimal pending-invitation projection. */
export function decodeLiveStudentCourseInvitations(
  value: unknown,
  path = "response",
): ReadonlyArray<LiveStudentCourseInvitationSummary> {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["invitations"]);
  return decodeArray(
    field(record, "invitations", path),
    `${path}.invitations`,
    decodeInvitationSummary,
  );
}

/** Rejects anything beyond the current Student's minimal Assessment landing projection. */
export function decodeLiveStudentAssessmentLandings(
  value: unknown,
  path = "response",
): ReadonlyArray<LiveStudentAssessmentLandingSummary> {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["assessments"]);
  return decodeArray(
    field(record, "assessments", path),
    `${path}.assessments`,
    decodeAssessmentSummary,
  );
}

/** Rejects any fields outside the authenticated Student's Course Progress projection. */
export function decodeStudentCourseProgress(
  value: unknown,
  path = "response",
): ReadonlyArray<StudentCourseProgressAssessment> {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["assessments"]);
  return decodeArray(
    field(record, "assessments", path),
    `${path}.assessments`,
    decodeProgressAssessment,
  );
}

/** Rejects values outside the safe fixed Coursework Active Attempt shortcut projection. */
export function decodeStudentCourseActiveAttempt(
  value: unknown,
  path = "response",
): StudentCourseActiveAttempt {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["assessmentAttemptId", "startedAt", "latestActivityAt"]);
  const rawAttemptId = decodeNullable(
    field(record, "assessmentAttemptId", path),
    `${path}.assessmentAttemptId`,
    decodeString,
  );
  const assessmentAttemptId = rawAttemptId === null ? null : parseAssessmentAttemptId(rawAttemptId);
  if (rawAttemptId !== null && assessmentAttemptId === null) {
    throw new DecodeError(`${path}.assessmentAttemptId`, "a canonical Assessment Attempt UUID");
  }
  const startedAt = decodeNullable(
    field(record, "startedAt", path),
    `${path}.startedAt`,
    decodeNonnegativeInteger,
  );
  const latestActivityAt = decodeNullable(
    field(record, "latestActivityAt", path),
    `${path}.latestActivityAt`,
    decodeNonnegativeInteger,
  );
  if (
    (assessmentAttemptId === null && (startedAt !== null || latestActivityAt !== null)) ||
    (assessmentAttemptId !== null && (startedAt === null || latestActivityAt === null))
  ) {
    throw new DecodeError(path, "an Attempt ID with both timestamps, or three explicit nulls");
  }
  return { assessmentAttemptId, startedAt, latestActivityAt };
}

/** Rejects values outside the signed-in Student's Latest Feedback shortcut projection. */
export function decodeStudentLatestFeedback(
  value: unknown,
  path = "response",
): StudentLatestFeedback {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["assessmentAttemptId"]);
  const rawAttemptId = decodeNullable(
    field(record, "assessmentAttemptId", path),
    `${path}.assessmentAttemptId`,
    decodeString,
  );
  const assessmentAttemptId = rawAttemptId === null ? null : parseAssessmentAttemptId(rawAttemptId);
  if (rawAttemptId !== null && assessmentAttemptId === null) {
    throw new DecodeError(`${path}.assessmentAttemptId`, "a canonical Assessment Attempt UUID");
  }
  return { assessmentAttemptId };
}

File: crates/learning-data-access/src/live_student_course_landing.rs
//! Student-authorized Course and released Assessment landing projections.
//!
//! The landing boundary exposes only public Course and Assessment IDs
//! with learner-facing titles. PostgreSQL derives the current active Student
//! Account and exact active Student Course Membership before it projects them.

use async_trait::async_trait;
use browser_api_contract::student_assessment_decision::StudentAssessmentDecisionSummary;
use question_model::{
    AssessmentAttemptCompletion, AssessmentAttemptId, AssessmentId, AssessmentType,
    CourseInstanceId, CourseTerm, PublishedQuestionRevisionTuple, Timestamp,
};
use serde::Serialize;

use crate::{SessionTokenHash, StoreError};

/// One point-based Assessment grade contribution selected across submitted Attempts.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveAssessmentGradeContribution {
    pub points_earned: f64,
    pub points_possible: f64,
}

/// One active Student Course Instance available from the landing page.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStudentCourseLandingSummary {
    /// Public Course Instance ID.
    pub course_instance_id: CourseInstanceId,
    /// Compact Course Instance name for constrained navigation.
    pub short_name: String,
    /// Descriptive Course Instance name for headings and lists.
    pub long_name: String,
}

/// One pending Student Course Invitation available from the landing page.
///
/// This remains distinct from [`LiveStudentCourseLandingSummary`]: a pending
/// Course Invitation is not an active Student Course Membership.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStudentCourseInvitationSummary {
    /// Public Course Instance ID, never an invitation or Account identity.
    pub course_instance_id: CourseInstanceId,
    /// Compact Course Instance name for constrained navigation.
    pub short_name: String,
    /// Descriptive Course Instance name for headings and lists.
    pub long_name: String,
    /// Server-verified display name of the Course's assigned Instructor.
    pub instructor_display_name: String,
    /// Inclusive Course calendar dates, available before accepting the invitation.
    pub term: CourseTerm,
}

/// One released Assessment available from an authorized Student Course.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStudentAssessmentLandingSummary {
    /// Public Assessment ID.
    pub assessment_id: AssessmentId,
    /// Student-facing released Assessment title.
    pub title: String,
    /// Product-defined pedagogical Type for this Assessment.
    pub assessment_type: AssessmentType,
    /// Same server-owned policy and start decision returned by Assessment Access.
    pub decision: StudentAssessmentDecisionSummary,
    /// One-based current Assessment Attempt number, or none before work starts.
    pub assessment_attempt_number: Option<u32>,
    /// Current Assessment Attempt completion, or none when work has not started.
    pub assessment_attempt_completion: Option<AssessmentAttemptCompletion>,
    /// Whether the ordinary start action would resume an existing Assessment Attempt.
    pub can_resume_assessment_attempt: bool,
    /// Questions with an immutable Grading Result in the current Assessment Attempt.
    pub graded_question_count: u32,
    /// Complete durable saved responses in the latest Assessment Attempt; not grading evidence.
    pub saved_question_count: u32,
    /// Total questions in the current Assessment; an existing Attempt retains its issued-question evidence.
    pub question_count: u32,
    /// Highest submitted Assessment score when that Attempt's disclosure permits it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assessment_score: Option<LiveAssessmentGradeContribution>,
}

/// One released Assessment's self-only Course Progress evidence.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStudentCourseProgressAssessment {
    pub assessment_id: AssessmentId,
    pub title: String,
    pub assessment_type: AssessmentType,
    pub assessment_attempt_count: u32,
    pub submitted_assessment_attempt_count: u32,
    pub latest_assessment_attempt_number: Option<u32>,
    pub latest_assessment_attempt_completion: Option<AssessmentAttemptCompletion>,
    pub latest_activity_at_millis: Option<i64>,
    /// Highest submitted Attempt score only when that Attempt's disclosure permits it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assessment_score: Option<LiveAssessmentGradeContribution>,
    pub assessment_score_is_latest_attempt: bool,
}

/// One self-only Attempt in a Course-wide, newest-first history page.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStudentCourseAttemptHistoryEntry {
    pub assessment_attempt_id: AssessmentAttemptId,
    pub assessment_id: AssessmentId,
    pub assessment_title: String,
    pub assessment_attempt_number: u32,
    pub started_at: Timestamp,
    pub submitted_at: Option<Timestamp>,
    /// This Attempt's score only when grading is complete and disclosure permits it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assessment_score: Option<LiveAssessmentGradeContribution>,
}

/// One exact immutable Published Question Revision's disclosed self-only outcomes.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStudentCoursePracticeQuestionStats {
    pub published_question_revision_tuple: PublishedQuestionRevisionTuple,
    pub full_credit_attempt_count: u64,
    pub partial_credit_attempt_count: u64,
    pub incorrect_attempt_count: u64,
    pub unanswered_attempt_count: u64,
    pub disclosed_attempt_count: u64,
    pub not_full_credit_count: u64,
    pub average_display_duration_ms: Option<f64>,
    pub display_duration_sample_count: u64,
    pub relevant_assessment_attempt_id: AssessmentAttemptId,
}

/// Session-authorized persistence boundary for the Student Course landing.
#[async_trait]
pub trait LiveStudentCourseLandingStore: Send + Sync {
    /// Lists only Course Instances with an active Student Course Membership.
    async fn list_live_student_courses(
        &self,
        session_token_hash: SessionTokenHash,
    ) -> Result<Vec<LiveStudentCourseLandingSummary>, StoreError>;

    /// Lists only the authenticated Student's pending, unexpired Course Invitations
    /// for Courses without an active Student Course Membership.
    ///
    /// The projection intentionally omits invitation, Account, Student Record,
    /// and membership identities. Accepting an invitation remains the separate
    /// exact-transaction Course Roster operation.
    async fn list_pending_live_student_course_invitations(
        &self,
        session_token_hash: SessionTokenHash,
    ) -> Result<Vec<LiveStudentCourseInvitationSummary>, StoreError>;

    /// Lists released Assessments after exact Course-membership authorization.
    ///
    /// A foreign, malformed, inactive, or ended Course is concealed as
    /// [`StoreError::Forbidden`]; an authorized Course without releases returns
    /// an empty list.
    async fn list_released_live_student_assessments(
        &self,
        session_token_hash: SessionTokenHash,
        course_instance_id: CourseInstanceId,
    ) -> Result<Vec<LiveStudentAssessmentLandingSummary>, StoreError>;

    /// Lists disclosed Progress and Attempt activity for one authorized Course.
    async fn list_live_student_course_progress(
        &self,
        session_token_hash: SessionTokenHash,
        course_instance_id: CourseInstanceId,
    ) -> Result<Vec<LiveStudentCourseProgressAssessment>, StoreError>;

    /// Lists the authenticated Student's Course Attempts by bounded keyset pages.
    async fn list_live_student_course_attempt_history(
        &self,
        session_token_hash: SessionTokenHash,
        course_instance_id: CourseInstanceId,
        page: crate::PageRequest,
    ) -> Result<crate::Page<LiveStudentCourseAttemptHistoryEntry>, StoreError>;

    /// Lists only the current Student's Course-scoped disclosed Question outcomes.
    async fn list_live_student_course_practice_stats(
        &self,
        session_token_hash: SessionTokenHash,
        course_instance_id: CourseInstanceId,
    ) -> Result<Vec<LiveStudentCoursePracticeQuestionStats>, StoreError>;
}
